### ISSUE TITLE
Homebridge detects no accessory 

### ISSUE THREAD

**Issue by nicolinuxfr** [loki29pl/homebridge-weconn#1]
First, let me thank you for this plugin ! 

I'm trying to use it for a WiFi Smart Plug that relies on the Weconn app. I think it's a generic plug, sold under the "Horsky" brand, at least in France ([Amazon](https://amzn.to/2uOCGxW)).

I followed the config sample to add the plug. I'm sure the IP and Mac address are correct. But when I launch homebridge, it detects 0 accessory : 

```
Apr  4 18:32:33 raspberry homebridge[13106]: [2018-4-4 18:32:33] Loaded plugin: homebridge-platform-wemo
Apr  4 18:32:33 raspberry homebridge[13106]: [2018-4-4 18:32:33] Registering platform 'homebridge-platform-wemo.BelkinWeMo'
Apr  4 18:32:33 raspberry homebridge[13106]: [2018-4-4 18:32:33] ---
Apr  4 18:32:34 raspberry homebridge[13106]: [2018-4-4 18:32:34] Loaded plugin: homebridge-tplink-smarthome
Apr  4 18:32:34 raspberry homebridge[13106]: [2018-4-4 18:32:34] Registering platform 'homebridge-tplink-smarthome.TplinkSmarthome'
Apr  4 18:32:34 raspberry homebridge[13106]: [2018-4-4 18:32:34] ---
Apr  4 18:32:34 raspberry homebridge[13106]: [2018-4-4 18:32:34] Loaded plugin: homebridge-weconn
Apr  4 18:32:34 raspberry homebridge[13106]: [2018-4-4 18:32:34] Registering accessory 'homebridge-weconn.WeConn'
Apr  4 18:32:34 raspberry homebridge[13106]: [2018-4-4 18:32:34] ---
Apr  4 18:32:34 raspberry homebridge[13106]: [2018-4-4 18:32:34] Loaded config.json with 0 accessories and 2 platforms.
```

Here's my config.json :

```
{
	"bridge": {
		"name": "Homebridge",
		"username": "XX:XX:XX:XX:XX:XX",
		"port": 51826,
		"pin": "XXX-XXX-XXX"
	},

	"platforms": [{
			"platform": "BelkinWeMo",
			"name": "WeMo Platform"
		},
		{
			"platform": "TplinkSmarthome",
			"name": "TplinkSmarthome"
		}
	],

	"accessories": [{
		"accessory": "WeConn",
		"name": "Lampes2",
		"ip_address": "192.168.8.133",
		"mac": "XX:XX:XX:XX:XX:XX"
	}]
}
```

One thing I don't know is the name parameter. I used the exact same name as in the mobile app, but maybe I'm supposed to use something else ?

Let me know if you need more information !

**Comment by loki29pl** [comment]
Your config looks ok. Which version of Homebridge and Node are you using?

**Comment by nicolinuxfr** [comment]
homebridge 0.4.38 and nodejs v9.10.1

**Comment by loki29pl** [comment]
I've added debug option. Please download new version and enable debug. I've updated readme. Try start this plugin only with accessories.

**Comment by nicolinuxfr** [comment]
Sorry, I was editing the wrong config.json… 🤦‍♂️

Anyway, after correcting my mistake, the accessory does appear in the logs and in my Home app on iOS. But the accessory is not available and when I try to use it, it crashes Homebridge with this error :

```
events.js:165
Apr  4 19:20:13 raspberry homebridge[1277]:       throw er; // Unhandled 'error' event
Apr  4 19:20:13 raspberry homebridge[1277]:       ^
Apr  4 19:20:13 raspberry homebridge[1277]: Error: connect ECONNREFUSED 192.168.8.133:9957
Apr  4 19:20:13 raspberry homebridge[1277]:     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1173:14)
Apr  4 19:20:13 raspberry homebridge[1277]: Emitted 'error' event at:
Apr  4 19:20:13 raspberry homebridge[1277]:     at emitErrorNT (internal/streams/destroy.js:64:8)
Apr  4 19:20:13 raspberry homebridge[1277]:     at process._tickCallback (internal/process/next_tick.js:114:19)
```

**Comment by loki29pl** [comment]
Check ip address, set to static in your DHCP configuration.

**Comment by nicolinuxfr** [comment]
The address is correct. I’ll try with static configuration but it’s not possible with my current router. 

Maybe it has nothing to do with the problem but why does my plug show as a switch in the iOS app ? (Lampes2 is the Weconn one)

**Comment by loki29pl** [comment]
You must set correct IP address of plug. You can try `telnet 192.168.8.133 9957`
It's shows up as a switch and then you can on/off your plug.

**Comment by nicolinuxfr** [comment]
It is the correct IP, I'm sure of that (the MAC address on my router matches the one inside Weconn app). And even if it is not static in my DHCP, the IP does not change, at least not until now.

Telnet fails, though :

```
➜  ~ telnet 192.168.8.133 9957
Trying 192.168.8.133...
telnet: connect to address 192.168.8.133: Connection refused
telnet: Unable to connect to remote host
```

Maybe it's a security issue with my plug ?

**Comment by nicolinuxfr** [comment]
Maybe it will help : I tried the `tcpdump` command from the [Stack Overflow answer you linked to](https://stackoverflow.com/questions/31827645/how-to-control-wifi-plug-gwf-s171using-raspberry-pi/31929689#31929689). The `grep` gave me nothing, so I tried without and looked for the IP of the plug. 

Here's what I got, if it can help you debug :

```
07:50:55.136333 IP (tos 0x0, ttl 128, id 8553, offset 0, flags [none], proto UDP (17), length 187)
    192.168.8.133.37261 > 255.255.255.255.20003: UDP, length 159
	0x0000:  ffff ffff ffff dc4f 2201 3bcd 0800 4500
	0x0010:  00bb 2169 0000 8011 4f9c c0a8 0885 ffff
	0x0020:  ffff 918d 4e23 00a7 3fe5 fef0 9f00 0305
	0x0030:  0200 0000 0000 fb05 0100 0000 b5b0 2700
	0x0040:  0000 3cb9 c55a 0000 0000 0000 0000 0000
	0x0050:  f0fe 0400 4c61 6d70 6573 3200 0000 0000
	0x0060:  0000 0000 0000 0000 0000 0000 0000 0000
	0x0070:  0000 0000 0105 00c0 a808 85dc 4f22 013b
	0x0080:  cd00 0000 0000 0000 0302 5354 506c 7567
	0x0090:  5f33 4243 4400 0000 0000 0000 0000 0000
	0x00a0:  0000 0000 0000 0000 0000 0204 0000 1500
	0x00b0:  0101 0000 0000 0000 0020 2f00 0000 0000
	0x00c0:  0000 0000 0049 c3be f8
```

**Comment by loki29pl** [comment]
If telnet on port 9957 fails, this plugin won't work.

**Comment by nicolinuxfr** [comment]
I understand, I was wondering if I could find if it needs some other port or something ?

**Comment by loki29pl** [comment]
Try scan open ports on the device. Are you sure, you have a connection beetwen homebridge device and plug, does the ping work?

**Comment by nicolinuxfr** [comment]
Yes, ping does work from the device, to the plug.

Does that help for the open ports part ?

```
sudo nmap -sS -O 192.168.8.133

Starting Nmap 7.40 ( https://nmap.org ) at 2018-04-05 12:26 CEST
Nmap scan report for ESP_013BCD (192.168.8.133)
Host is up (0.0020s latency).
Not shown: 999 closed ports
PORT      STATE SERVICE
10000/tcp open  snet-sensor-mgmt
MAC Address: DC:4F:22:01:3B:CD (Unknown)
No exact OS matches for host (If you know what OS is running on it, see https://nmap.org/submit/ ).
TCP/IP fingerprint:
OS:SCAN(V=7.40%E=4%D=4/5%OT=10000%CT=1%CU=33639%PV=Y%DS=1%DC=D%G=Y%M=DC4F22
OS:%TM=5AC5FA01%P=arm-unknown-linux-gnueabihf)SEQ(SP=4E%GCD=1%ISR=AB%TI=I%C
OS:I=I%TS=U)OPS(O1=M5B4%O2=M5B4%O3=M5B4%O4=M5B4%O5=M5B4%O6=M5B4)WIN(W1=16D0
OS:%W2=16D0%W3=16D0%W4=16D0%W5=16D0%W6=16D0)ECN(R=Y%DF=N%T=80%W=16D0%O=M5B4
OS:%CC=N%Q=)T1(R=Y%DF=N%T=80%S=O%A=S+%F=AS%RD=0%Q=)T2(R=N)T3(R=Y%DF=N%T=80%
OS:W=16D0%S=O%A=S+%F=AS%O=M5B4%RD=0%Q=)T4(R=Y%DF=N%T=80%W=16D0%S=A+%A=S%F=A
OS:R%O=%RD=0%Q=)T5(R=Y%DF=N%T=80%W=16D0%S=A%A=S+%F=AR%O=%RD=0%Q=)T6(R=Y%DF=
OS:N%T=80%W=16D0%S=A%A=S%F=AR%O=%RD=0%Q=)T7(R=Y%DF=N%T=80%W=16D0%S=A%A=S+%F
OS:=AR%O=%RD=0%Q=)U1(R=Y%DF=N%T=80%IPL=38%UN=0%RIPL=G%RID=G%RIPCK=G%RUCK=G%
OS:RUD=G)IE(R=Y%DFI=S%T=80%CD=S)

Network Distance: 1 hop

OS detection performed. Please report any incorrect results at https://nmap.org/submit/ .
Nmap done: 1 IP address (1 host up) scanned in 21.16 seconds
```

**Comment by loki29pl** [comment]
I think you have a connection problem or your device is something different.

**Comment by nicolinuxfr** [comment]
I guess so, but since it's using the same weconn app…

**Comment by nicolinuxfr** [comment]
I just bought a new plug from a different brand that uses the WeConn app and it does not work either.

Do I have absolutely no luck at all or is there a problem with the plugin ?

**Comment by loki29pl** [comment]
Can you telnet plug on port 9957?

**Comment by nicolinuxfr** [comment]
Nope, it's the same thing as before. 

I'd like to help discovering why, but I don't know where to begin…

**Comment by DJM0** [comment]
@nicolinuxfr so I have a Weconn plug that doesn't work with this project either ([amazon](https://www.amazon.co.uk/gp/product/B076HL423Q)). It works fine with the Weconn Plug app, but just like you can see from your nmap scan it's using `tcp/10000` and not `tcp/9957`. So I suspect ours is either a newer or older version of the hardware @loki29pl wrote this for.

At first I tried sending the same packets to `tcp/10000` hoping they just changed the port, but after spending a few hours trying to decode network captures to the plug it looks like this implementation is a lot more complicated.

I believe this is the packet/data which switches the plug on:

<img width="901" alt="screen shot 2018-05-25 at 20 18 40" src="https://user-images.githubusercontent.com/1108606/40562434-dee52992-6058-11e8-9421-7b830768719e.png">

But certain bytes change every time I send the request, and I haven't worked out if it's an incrementing counter, time data or a challenge/response type security feature yet.

I'm planning to look into it a bit more this weekend. Looking inside the unit it has accessible debugging pins, so it might just be easier to re-flash it with one of the many esp8266 smart plug projects instead.

**Comment by ssamjh** [comment]
@davidmaitland I also have a device that is using port 10000/tcp.

Did you ever have much luck finding out the pattern or reverse engineering it?

**Comment by DJM0** [comment]
@ssamjh I spent some time on it, but didn't get anywhere as I think it is using some sort of encryption / security features. I'm not very experienced with reviser engineering so might be missing something simple.

Figured it would be easier to flash the ESP directly with custom firmware.

While doing the Wireshark I found the stock firmware sends a lot of requests to remote servers. I also found their user API for the app exposes personal data (emails, numbers, addresses) for every single registered user. So there is no way I'm running that on my home network 😆
